### PR TITLE
feat(lua): add `vim.fs.abspath`

### PR DESCRIPTION
**Pull request by famiu**
Problem: There is currently no way to check if a given path is absolute or convert a relative path to an absolute path through the Lua stdlib. `vim.fs.joinpath` does not work when the path is absolute. There is also currently no way to resolve `C:foo\bar` style paths in Windows.

Solution: Add `vim.fs.abspath`, which allows converting any path to an absolute path. This also allows checking if current path is absolute by doing `vim.fs.abspath(path) == path`. It also has support for `C:foo\bar` style paths in Windows.

### CODE REVIEW

File: runtime/lua/vim/fs.lua
Comment (lewis6991) [inline runtime/lua/vim/fs.lua:96]:
```suggestion
--- Current working directory to use as the base for the relative path.
--- This option is also ignored for C:foo\bar style paths on Windows, as those paths use the current
--- directory of the drive specified in the path.
--- (default: |current-directory|)
--- @field cwd? string
```

Should produce nicer docs.

File: runtime/lua/vim/fs.lua
Comment (lewis6991) [inline runtime/lua/vim/fs.lua:114]: Move this to below line 134

**Review by lewis6991 — APPROVED** [review]

File: runtime/doc/lua.txt
Comment (dundargoc) [inline runtime/doc/lua.txt:2859]: Yeah I still don't get the point of this. I can't come up with a reason you'd need this. If you need to know the absolute path from a file that isn't relative to the current working directory, does that not mean that you already know the absolute path of the file as it would be `cwd + path`? If so, why can't this just be done with string concatenation?

Comment (famiu) [reply runtime/doc/lua.txt:2859]: > Yeah I still don't get the point of this. I can't come up with a reason you'd need this. If you need to know the absolute path from a file that isn't relative to the current working directory, does that not mean that you already know the absolute path of the file as it would be `cwd + path`? If so, why can't this just be done with string concatenation?

Not necessarily. Assume you have a `cwd`, but you don't know if `path` is absolute or relative. Doing `cwd + path` won't work if `path` is already absolute. In which case, you'd first have to do a check if `path` is absolute by doing `abspath(path) == path`, and _then_ do `cwd + path` if path is not absolute, which imo is way too complicated to expect the average Neovim user to understand.

Comment (dundargoc) [reply runtime/doc/lua.txt:2859]: > Assume you have a `cwd`

I already have a problem with this premise. Why would I have a random `cwd` that isn't the current working directory?

Comment (famiu) [reply runtime/doc/lua.txt:2859]: > > Assume you have a `cwd`
> 
> I already have a problem with this premise. Why would I have a random `cwd` that isn't the current working directory?

I'll give you a usecase that I recently had, where I made a `Rename` command that allowed me to rename any file, and the new name would actually be a path that's relative to the directory of the file being renamed. Another usecase I can think of would be a fuzzy file searcher that allows you to specify a directory, then shows you the filenames relative to the target directory you specified.

I'm sure @echasnovski can think of other usecases as well.

Comment (dundargoc) [reply runtime/doc/lua.txt:2859]: > I'll give you a usecase that I recently had, where I made a Rename command that allowed me to rename any file, and the new name would actually be a path that's relative to the directory of the file being renamed.

Can I take a look at the code? Could you link it?

Comment (famiu) [reply runtime/doc/lua.txt:2859]: > > I'll give you a usecase that I recently had, where I made a Rename command that allowed me to rename any file, and the new name would actually be a path that's relative to the directory of the file being renamed.
> 
> Can I take a look at the code? Could you link it?

https://github.com/famiu/dot-nvim/blob/master/plugin/file_cmds.lua

Comment (echasnovski) [reply runtime/doc/lua.txt:2859]: > I'm sure @echasnovski can think of other usecases as well, considering that he agreed with having `cwd` in Matrix.

Sure. One of the real examples for me was to get absolute paths of particular subdirectories for all directories in 'runtimepath'. Here we have "many `cwd`" + "single `path`".

Now, this particular case (and many more) can be solved with `vim.fs.joinpath(vim.fs.abspath(cwd), path)`, but it is not 100% the case due to some obscure Windwos paths. So having `opts.cwd` seems logical to me.

Comment (famiu) [reply runtime/doc/lua.txt:2859]: > > I'm sure @echasnovski can think of other usecases as well, considering that he agreed with having `cwd` in Matrix.
> 
> Sure. One of the real examples for me was to get absolute paths of particular subdirectories for all directories in 'runtimepath'. Here we have "many `cwd`" + "single `path`".
> 
> Now, this particular case (and many more) can be solved with `vim.fs.joinpath(vim.fs.abspath(cwd), path)`, but it is not 100% the case due to some obscure Windwos paths. So having `opts.cwd` seems logical to me.

It also just adds 3 lines of code and is very low-cost, so basically not much reason to not have it.

Comment (gpanders) [reply runtime/doc/lua.txt:2859]: Specifying a separate `cwd` is fine. This is similar to the `openat` function. It's not a common use case, but there's no reason not to support it (especially if it's easy to support, as @famiu said).

File: runtime/lua/vim/fs.lua
Comment (justinmk) [inline runtime/lua/vim/fs.lua:129]: could be a ternary, which also avoids the ned for a `@type` annotation

Comment (famiu) [reply runtime/lua/vim/fs.lua:129]: Actually, this entire code can be removed once #28203 is merged, since that allows us a much simpler way of detecting absolute paths

File: runtime/lua/vim/fs.lua
Comment (justinmk) [inline runtime/lua/vim/fs.lua:105]: if `paths` is a list, that has a few advantages:

- allows multiple segments to be combined, similar to nodejs resolve() https://nodejs.org/api/path.html#pathresolvepaths
- when calling, the "cwd" is the first arg, which is more readable and more explicit. compare:
    - `abspath('foo', {cwd: '.'})`
    - `abspath({'.', 'foo'})`
    - making it explicit has other advantages, such as reminder plugin authors to think about `getcwd()` and window-local paths.

Comment (famiu) [reply runtime/lua/vim/fs.lua:105]: > * allows multiple segments to be combined, similar to nodejs resolve()

`vim.fs.joinpath()` can do that already

> * when calling, the "cwd" is the first arg, which is more readable and more explicit. compare:when calling, the "cwd" is the first arg, which is more readable and more explicit. compare:

Tbh with you, I personally do not find it more intuitive, the opposite, even. And even if it were more intuitive, in the vast majority of cases, you're going to not set `cwd` manually, and in those cases, just having a single path argument is much simpler.

Comment (justinmk) [reply runtime/lua/vim/fs.lua:105]: Alternatively could define it as `path: string|table`. That allows opt-in to explicit multi-arg form. Although that's a bit redundant with joinpath(), it eliminates the `opts.cwd` option.

> in the vast majority of cases, you're going to not set cwd manually, and in those cases, just having a single path argument is much simpler.

For the single-arg case we can default to using CWD as the root dir.

Comment (famiu) [reply runtime/lua/vim/fs.lua]: > Alternatively could define it as `path: string|table`. That allows opt-in to explicit multi-arg form. Although that's a bit redundant with joinpath(), it eliminates the `opts.cwd` option.
> 
> > in the vast majority of cases, you're going to not set cwd manually, and in those cases, just having a single path argument is much simpler.
> 
> For the single-arg case we can default to using CWD as the root dir.

We might need to add more opts (like path OS style and whether to expand env variables) to `abspath` in the future, so imo keeping the opts dict is best for now, it's also more in line with the abspath function in something like Python.  Having a dict just adds more complexity for the majority of usecases

Comment (justinmk) [reply runtime/lua/vim/fs.lua]: > We might need to add more opts (like path OS style and whether to expand env variables) to `abspath` in the future,

Irrelevant. The door is open for that in the future. But today, we can support a list arg as the first argument, that's natural and more readable than a CWD arg that _follows_ the first arg.

Comment (lewis6991) [reply runtime/lua/vim/fs.lua]: I'm not sure why we need a list argument when the user can simply use `vim.fs.joinpath`, especially since the use case here is already recognised to be quite rare.

Comment (justinmk) [reply runtime/lua/vim/fs.lua]: > I'm not sure why we need a list argument when the user can simply use `vim.fs.joinpath`

Because it avoids the `cwd` option.

Comment (famiu) [reply runtime/lua/vim/fs.lua]: > > I'm not sure why we need a list argument when the user can simply use `vim.fs.joinpath`
> 
> Because it avoids the `cwd` option. If we can avoid both, then that's fine too.

I would personally pick not having `cwd` over using a list argument. The list argument imo is not really intuitive. Here's how I see it:

`abspath` is a function that takes a path and returns the absolute path. What is the absolute path for a table supposed to mean? Imo it can really trip up new users. If removing `cwd` is the solution, then I'm fine with doing that

Comment (clason) [reply runtime/lua/vim/fs.lua]: For the record, I'm with famiu here. I don't understand the objection to the original signature at all, _provided_ that there's a compelling use case for resolving paths relative to non-current directories. (Maybe `cwd` is the wrong name and `base` or similar would be better?) If we currently don't have such a use case -- either in our own code or a reasonably common plugin -- we can leave it for later (with the understanding that we'll simply add it and _not_ start fighting over it again).

Comment (justinmk) [reply runtime/lua/vim/fs.lua]: Should we rather implement this as an option on `normalize()`? Not sure that `abspath()` is worth all the extra docs.

Comment (clason) [reply runtime/lua/vim/fs.lua]: Apples and oranges. `normalize()` does not need a cwd, since its action doesn't depend on it. Only the conversion relative to absolute path needs to know what the path is relative _to_ .

The thing to keep in mind is that you may wish to normalize a relative path, or make a non-normalized path absolute. So it's either small, composable, single-purpose functions (the *ooonix way*) or one big swiss army knife function that transforms paths, controlled by (a bunch of) options you need to track interactions of.

Comment (justinmk) [reply runtime/lua/vim/fs.lua]: > The thing to keep in mind is that you may wish to normalize a relative path, or make a non-normalized path absolute. 

What I meant was, similar to the `expand_env` option, normalize() could have a `abs` option. 

> So it's either small, composable, single-purpose functions (the _ooonix way_) or one big swiss army knife function that transforms paths, controlled by (a bunch of) options you need to track interactions of.

Yes, but the 2 main concepts are "normalize" and "resolve". Currently `normalize()` is also doing some "resolving" (the `expand_env` option), and arguably that should be pulled into a separate `resolve()` or `expand()` function.

Comment (justinmk) [reply runtime/lua/vim/fs.lua]: Anyway to be clear, this PR is not blocked on the above.

**Comment by justinmk** [comment]
Do we also want `relpath()` ? https://nodejs.org/api/path.html#pathrelativefrom-to

File: runtime/doc/lua.txt
Comment (justinmk) [inline runtime/doc/lua.txt:2867]: we can drop this for now

File: src/nvim/lua/executor.c
Comment (justinmk) [inline src/nvim/lua/executor.c:619]: the other calls above are inlined, do we need the separate `nlua_add_fs` function?

File: src/nvim/lua/fs.c
Comment (justinmk) [inline src/nvim/lua/fs.c:23]: do we not already have some internal C function that does this? is there any opportunity for deduplicating existing code?

File: test/functional/lua/fs_spec.lua
Comment (justinmk) [inline test/functional/lua/fs_spec.lua:430]: (untested)
```suggestion
    local cwd = assert(is_os('win') and (vim.uv.cwd()):gsub('\\', '/') or vim.uv.cwd())
    local home = assert(is_os('win') and (vim.uv.os_homedir()):gsub('\\', '/') or vim.uv.os_homedir())
```

File: test/functional/lua/fs_spec.lua
Comment (justinmk) [inline test/functional/lua/fs_spec.lua:448]: "correctly" and "works with..." are always redundant.
```suggestion
    it('expands ~', function()
```

File: src/nvim/lua/fs.c
Comment (famiu) [reply src/nvim/lua/fs.c:23]: None that I'm aware of

File: src/nvim/lua/executor.c
Comment (famiu) [reply src/nvim/lua/executor.c:619]: Not really, `nlua_add_fs` was created when the call was in `nlua_state_init`. Now that it's in `nlua_common_vim_init`, it's better to inline it

File: test/functional/lua/fs_spec.lua
Comment (famiu) [reply test/functional/lua/fs_spec.lua:430]: NOTE: cwd and home can be nil, but if we want to not consider that then that's fine

Comment (famiu) [reply test/functional/lua/fs_spec.lua:430]: If `cwd` or `home` are `nil`, `gsub` will not work

File: src/nvim/lua/fs.c
Comment (justinmk) [inline src/nvim/lua/fs.c:16]: Any chance `vim.uv.fs_realpath` avoids the need for this?

File: test/functional/lua/fs_spec.lua
Comment (justinmk) [reply test/functional/lua/fs_spec.lua:430]: > If `cwd` or `home` are `nil`, `gsub` will not work

it will error, just like the assert(). for the purposes of this test code that's pretty equivalent. though I don't know of LuaLS will like it.

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/20976 proposed `realpath`. How does it compare to `abspath` ? Do they have overlap? If so, maybe a `resolve()` or something like that makes more sense.

**Comment by echasnovski** [comment]
The `realpath()` works only if path is present on disk and resolves symlinks. The `vim.fs.abspath()` was intended with the sole purpose of transforming *any* path to be absolute (whatever it means depending on the OS) without unnecessary extra changes to the path itself.

**Comment by justinmk** [comment]
> The `vim.fs.abspath()` was intended with the sole purpose of transforming _any_ path to be absolute (whatever it means depending on the OS) without unnecessary extra changes to the path itself.

Doesn't `fslua_get_drive_cwd` kind of violate that? Because it is context-dependent.

We could have a flag that controlled the behavior. Main point is that `realpath` and `abspath` seem pretty related so having two different functions is confusing.

**Comment by famiu** [comment]
> > The `vim.fs.abspath()` was intended with the sole purpose of transforming _any_ path to be absolute (whatever it means depending on the OS) without unnecessary extra changes to the path itself.
> 
> Doesn't `fslua_get_drive_cwd` kind of violate that? Because it is context-dependent.

As context-dependent as it needs to be, not any more than that. Resolving symlinks when it's not intended goes a step beyond.

> We could have a flag that controlled the behavior. Main point is that `realpath` and `abspath` seem pretty related so having two different functions is confusing.

From what I understand, `realpath` is for resolving symlinks for existing paths, `abspath` is for converting relative paths to absolute paths, two very different usecases

File: src/nvim/lua/fs.c
Comment (famiu) [reply src/nvim/lua/fs.c:16]: Oh hey that actually does work. Thanks a lot for the suggestion

File: test/functional/lua/fs_spec.lua
Comment (famiu) [reply test/functional/lua/fs_spec.lua:430]: LuaLS does complain about it if I change it to that, it also looks very cluttered and hard to parse mentally, imo.

Comment (justinmk) [reply test/functional/lua/fs_spec.lua:430]: How is following a bunch of reassignments easier to parse mentally?

Ternaries are preferred. Not going to argue about this over and over.

File: runtime/doc/lua.txt
Comment (justinmk) [inline runtime/doc/lua.txt:2950]: > A tilde (~) character at the beginning of the path is expanded to the user's home directory.

That seems random, and somewhat against the spirit of "Does not check if the path exists".

Expanding the path should perhaps be delegated to [realpath()](https://github.com/neovim/neovim/pull/20976).

Comment (justinmk) [reply runtime/doc/lua.txt:2950]: Though this raises a question about whether `~foo` should also be supported.

File: test/functional/lua/fs_spec.lua
Comment (justinmk) [inline test/functional/lua/fs_spec.lua:460]: `fix_slashes` https://github.com/neovim/neovim/blob/487c48ec8689b865bad04fdb87b61f5ada25da97/test/testutil.lua#L50

File: runtime/doc/lua.txt
Comment (echasnovski) [reply runtime/doc/lua.txt:2950]: > That seems random, and somewhat against the spirit of "Does not check if the path exists".

I don't see how it is against it. It doesn't require checking if the path exist, only do some string substitution using OS data. And it does so because it is a common way to denote path starting from user's home directory.

> Though this raises a question about whether ~foo should also be supported.

That arguably should be '~foo' relative path. Perhaps wording in the description could be better with something like "A tilde (~) directory at the beginning of the path ...".

Comment (famiu) [reply runtime/doc/lua.txt:2950]: > Though this raises a question about whether `~foo` should also be supported.

Could be added later if we want. Both `vim.fs.normalize` and `vim.fs.abspath` use `expand_home` so that function could be modified to add that functionality to both of them

Comment (justinmk) [reply runtime/doc/lua.txt:2950]: > I don't see how it is against it. It doesn't require checking if the path exist, only do some string substitution using OS data.

It implies the existence of a filepath with a home-dir concept, instead of being purely "path only".

Comment (echasnovski) [reply runtime/doc/lua.txt:2950]: It does not look like it. It is the same concept as '.' and '..' - convention for a specific directory 

The reasoning of why '\~' is replaced and '.' / '..' are not seems sound: the '\~' denotes "absolute" path while '.' / '..' - relative. Not resolving tilde results in different (pure) path then the input, while not resolving '.' / '..' results in the same path. Having them "squashed" can be done by extra `vim.fs.normalize()` call.

Also, not resolving tilde would require something like `vim.fs.normalize(vim.fs.abspath(vim.fs.normalize(path)))` to ensure that the final result is fully resolved absolute path.

Comment (dundargoc) [reply runtime/doc/lua.txt:2950]: > For reference, [python's abspath](https://docs.python.org/3/library/os.path.html#os.path.abspath) doesn't expand "~", it does something braindead:

Python actually does the technically correct thing since you can create a directory named `~`. I guess you can argue if it's more beneficial to denote `~` to mean home directory rather than following all allowable paths.

Comment (echasnovski) [reply runtime/doc/lua.txt:2950]: That's a good point. I think that behavior of whether first '~' directory is treated as user's home should then be consistent across all `vim.fs`. For example, `normalize()` resolves it. So ideally they all do or do not resolve by default, while optionally provide a flag to control it. I'd say resolving by default is vastly more useful.

Comment (dundargoc) [reply runtime/doc/lua.txt:2950]: I'm unsure at the moment. Optionally supporting expansion for all `vim.fs` functions would mean adding an optional flag for all `vim.fs` function, and I suspect this could massively increase complexity/surface area of the `vim.fs` module. Instead, letting the user resolve the path once via `expand()` or a potential `vim.fs.realpath()` sounds like a simpler method.

Comment (dundargoc) [reply runtime/doc/lua.txt:2950]: I don't see why `vim.fs.normalize(vim.fs.abspath(vim.fs.normalize(path)))` is needed.

```lua
local path = "~/programs"
vim.print(vim.fs.normalize(path))
-- /Users/username/programs
```

I might be missing an edge case you found.

Comment (echasnovski) [reply runtime/doc/lua.txt:2950]: This isn't about any edge case per se. Assume the plugin gets from the user some `path` and needs to convert it to absolute fully resolved path. As good plugin does, it might want to allow user to at least be able to:
- Use '~/programs' to mean '/home/user/programs'.
- Use '../programs' to mean 'path/to/programs' (with current directory being 'path/to/cwd').

How plugin is expected to achieve that using only `vim.fs` in case `vim.fs.abspath()` doesn't resolve '\~'?
- Using *just* `vim.fs.normalize(path)` doesn't work for '../programs', as it returns '../programs'. But yes, does work for '~/programs' because `vim.fs.normalize()` resolves it (which won't be the case if "... letting the user resolve the path once via expand() or a potential vim.fs.realpath() sounds like a simpler method.")
- Using `vim.fs.abspath(vim.fs.normalize(path))` also doesn't work for '../programs' because it returns '/path/to/cwd/../programs', i.e. not a resolved path.
- Thus it needs to resort to `vim.fs.normalize(vim.fs.abspath(vim.fs.normalize(path)))` or a custom "expand only home directory" function instead of the inner `vim.fs.normalize()`. Which is, of course, pretty straightforward, but a bit annoying.

Comment (dundargoc) [reply runtime/doc/lua.txt:2950]: Wouldn't a potential `realpath` function alleviate this problem? Basically a function that says "give me the actual, for real path" where we can blast the path with any amount of path normalization and absolute pathism?

Comment (echasnovski) [reply runtime/doc/lua.txt:2950]: > Wouldn't a potential `realpath` function alleviate this problem? Basically a function that says "give me the actual, for real path" where we can blast the path with any amount of path normalization and absolute pathism?

I would assume any function named `realpath()` to work only on "actual, real paths", i.e. paths that are present on disk. This is not always helpful, like for example in cases like "create a file/directory based on the user input `path`" or "act based on user `path` which might or might not actually exist on disk".

Having the ability to work on "pure paths" (i.e. just strings that behave by the rules of file system paths) is pretty liberating as a plugin author. I think it should be at least an active aspiration for `vim.fs` to try and not rely on path actually be present on disk if it is possible to do so.

File: test/functional/lua/fs_spec.lua
Comment (dundargoc) [reply test/functional/lua/fs_spec.lua:460]: Fix in https://github.com/neovim/neovim/pull/31790

Comment (lewis6991) [reply test/functional/lua/fs_spec.lua:460]: I have no idea why that function accepts a table. We definitely shouldn't port that aspect over, nor do I think we need to expose such a function.

Comment (justinmk) [reply test/functional/lua/fs_spec.lua:460]: It's just a util function for tests. Tests should use it instead of doing it manually.

Comment (lewis6991) [reply test/functional/lua/fs_spec.lua:460]: But does that mean we have to expose this as stable API? It's just a basic gsub. Doesn't seem like the tradeoffs are being properly considered.

Comment (justinmk) [reply test/functional/lua/fs_spec.lua:460]: `fix_slashes` isn't being exposed (AFAIK).

are you commenting on `abspath` ?

Comment (dundargoc) [reply test/functional/lua/fs_spec.lua:460]: Yeah, what is bro on about

Comment (dundargoc) [reply test/functional/lua/fs_spec.lua:460]: Not exposing `fix_slashes` in my PR.

Comment (lewis6991) [reply test/functional/lua/fs_spec.lua:460]: I'm commenting specifically about `fix_slashes`, as mentioned in the first comment here.

Comment (dundargoc) [reply test/functional/lua/fs_spec.lua:460]: ngl, I'm kinda confused as well. We're not exposing or porting `fix_slashes`, it is only used in tests.

Comment (lewis6991) [reply test/functional/lua/fs_spec.lua:460]: Yeah sorry, I'm going to finger point GitHub mobile for this. Backing out of the notification and viewing the full code, gave me the context I was missing, mainly the filename and that this was a test. Sorry for the misunderstanding.

**Pull request merged**
